lib/Controller/SMimeCertificatesController.php

```diff
@@ -0,0 +1,141 @@
+<?php
+
+declare(strict_types=1);
+
+/**
+ * @copyright Copyright (c) 2022 Richard Steinmetz <richard@steinmetz.cloud>
+ *
+ * @author Richard Steinmetz <richard@steinmetz.cloud>
+ *
+ * @license AGPL-3.0-or-later
+ *
+ * This code is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License, version 3,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License, version 3,
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>
+ *
+ */
+
+namespace OCA\Mail\Controller;
+
+use Exception;
+use OCA\Mail\Db\SMimeCertificate;
+use OCA\Mail\Exception\ServiceException;
+use OCA\Mail\Http\JsonResponse;
+use OCA\Mail\Service\Attachment\UploadedFile;
+use OCA\Mail\Service\SMimeService;
+use OCP\AppFramework\Controller;
+use OCP\AppFramework\Db\DoesNotExistException;
+use OCP\AppFramework\Http;
+use OCP\IRequest;
+use OCP\Security\ICrypto;
+
+class SMimeCertificatesController extends Controller {
+	private ?string $userId;
+	private SMimeService $certificateService;
+	private ICrypto $crypto;
+
+	public function __construct(string $appName,
+								IRequest $request,
+								?string $userId,
+								SMimeService $certificateService,
+								ICrypto $crypto) {
+		parent::__construct($appName, $request);
+		$this->userId = $userId;
+		$this->certificateService = $certificateService;
+		$this->crypto = $crypto;
+	}
+
+	/**
+	 * @NoAdminRequired
+	 * @TrapError
+	 *
+	 * @throws ServiceException
+	 */
+	public function index(): JsonResponse {
+		$certificates = $this->certificateService->findAllCertificates($this->userId);
+		$certificates = array_map(function (SMimeCertificate $certificate) {
+			try {
+				$decryptedCertificate = $this->crypto->decrypt($certificate->getCertificate());
+			} catch (Exception $e) {
+				throw new ServiceException(
+					'Failed to decrypt certificate: ' . $e->getMessage(),
+					0,
+					$e,
+				);
+			}
+			$json = $certificate->jsonSerialize();
+			$json['info'] = $this->certificateService->parseCertificate($decryptedCertificate);
+			return $json;
+		}, $certificates);
+		return JsonResponse::success($certificates);
+	}
+
+	/**
+	 * @NoAdminRequired
+	 * @TrapError
+	 *
+	 * @param int $id
+	 * @return JsonResponse
+	 *
+	 * @throws DoesNotExistException
+	 */
+	public function destroy(int $id): JsonResponse {
+		$this->certificateService->deleteCertificate($id, $this->userId);
+		return JsonResponse::success();
+	}
+
+	/**
+	 * @NoAdminRequired
+	 * @TrapError
+	 *
+	 * @return JsonResponse
+	 *
+	 * @throws ServiceException
+	 */
+	public function create(): JsonResponse {
+		// TODO: What about PKCS12 certificates?
```

Comment:
yep just what I ran into while testing. it would make sense to note this on the UI

Reply:
Something like:

```
Only PEM encoded certificates and private keys are supported. PKCS #12 certificates (.p12) can't be imported and need to be converted.
```

Or is this too technical?

Reply:
:+1: I think it's hard not to get technical here